### PR TITLE
Call typeProcessingOver even if an error occurred

### PR DESCRIPTION
**Pull request by wmdietl**
Possible changes to #3136
Still needs some documentation tweaks, but I need to run now.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/common/util/count/AnnotationStatistics.java [hunk 1]
```diff
@@ -66,6 +67,12 @@
 
     @Override
     public void typeProcessingOver() {
+        Log log = getCompilerLog();
+        if (log.nerrors != 0) {
+            System.out.printf(
+                    "Compilation issued %d error(s). Report might be partial.%n", log.nerrors);
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/common/util/count/AnnotationStatistics.java]: This is easy for a user to overlook.  I think it's better to produce no output than potentially incorrect output.  Same comment elsewhere.

File: framework/src/main/java/org/checkerframework/common/basetype/BaseTypeChecker.java [hunk 2]
```diff
@@ -695,7 +695,6 @@ public void typeProcessingOver() {
         for (BaseTypeChecker checker : getSubcheckers()) {
             checker.typeProcessingOver();
         }
-
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/common/basetype/BaseTypeChecker.java]: Nit: spacing change.

File: javacutil/src/main/java/org/checkerframework/javacutil/AbstractTypeProcessor.java [hunk 3]
```diff
@@ -139,11 +139,20 @@ public void typeProcessingStart() {}
...
      *
-     * <p>If an error (a Java error or a processor error) is reported, this method is not guaranteed
-     * to be invoked.
+     * @param hasError true if compilation issued an error, either from the Java compiler or from a
+     *     pluggable type-checker
      */
     public void typeProcessingOver() {}
 
+    /**
+     * Return log.
```
Comment (mernst) [inline javacutil/src/main/java/org/checkerframework/javacutil/AbstractTypeProcessor.java]: Could you make this more descriptive?  A "log" is customarily unstructured text, but this has structure as explained at the com.sun.tools.javac.util.Log javadoc.

File: javacutil/src/main/java/org/checkerframework/javacutil/AbstractTypeProcessor.java [hunk 4]
```diff
@@ -139,11 +139,20 @@ public void typeProcessingStart() {}
      * <p>Subclasses may override this method to do any aggregate analysis (e.g. generate report,
      * persistence) or resource deallocation.
      *
-     * <p>If an error (a Java error or a processor error) is reported, this method is not guaranteed
-     * to be invoked.
+     * @param hasError true if compilation issued an error, either from the Java compiler or from a
```
Comment (mernst) [inline javacutil/src/main/java/org/checkerframework/javacutil/AbstractTypeProcessor.java]: You don't want this change.
But do change to say that the method is guaranteed to be called.
The Javadoc should probably contain a hint about getCompilerLog.nerrors.

**Comment by wmdietl** [comment]
Will be finished in #3136.

**Pull request closed without merging**